src/persistQueryClient/PersistQueryClientProvider.tsx

```diff
@@ -0,0 +1,44 @@
+import React from 'react'
+
+import { persistQueryClient, PersistQueryClientOptions } from './persist'
+import {
+  QueryClientProvider,
+  QueryClientProviderProps,
+  IsHydratingProvider,
+} from '../reactjs'
+
+export interface PersistQueryClientProviderProps
+  extends QueryClientProviderProps {
+  persistOptions: Omit<PersistQueryClientOptions, 'queryClient'>
+}
+
+export const PersistQueryClientProvider = ({
+  client,
+  children,
+  persistOptions,
+  ...props
+}: PersistQueryClientProviderProps): JSX.Element => {
+  const [isHydrating, setIsHydrating] = React.useState(true)
+  const options = React.useRef(persistOptions)
+  React.useEffect(() => {
+    let unsubscribe: (() => void) | undefined
+
+    async function run() {
+      unsubscribe = await persistQueryClient({
```

Comment:
I think this might have a potential (theoretical?) issue. You don't get back the unsubscribe callback until after having awaited the `persistQueryClient`-call. If the effect cleans up while that is going on, it will never call unsubscribe and you will have a hanging subscription (not sure what consequences that will have).

Reply:
yes, that's true. `persistQueryClient` is implemented as:

```
await persistQueryClientRestore(props)
return persistQueryClientSubscribe(props)
```

subscribing is sync, but restoring is async. In order to `await` it, we must return a promise. We could subscribe _before_ restoring, but I have bigger doubts about that because the restore itself might trigger a write then. Do you see a good way to solve this?

Reply:
I'm not sure it would be an issue in reality, but, maybe a "solution" could be to return an unsubscribe and a Promise synchronously. If the unsubscribe is called before the subscription happened, we just keep track of that fact and don't subscribe when the time comes. After subscription happens, unsubscribe just does its thing?

Reply:
need to add some tests to see if it works, but is this what you had in mind?

261777cc

Reply:
Left a small comment on a detail, but yes, exactly the approach I was thinking about! 😃

Reply:
good catch. next attempt: 0f6de7a7